File: problem5.rb
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
# 
# What is the smallest number that is evenly divisible by all of the numbers from 1 to 20?

def gcd(x,y)
	return gcd(y,x) if x<y
	return y if x % y == 0
	return gcd(y, x % y)
end

def lcm(x,y)
	return x*y / gcd(x,y)
end

puts (2..20).inject(1){|x,y| lcm(x,y)}

File: problem5.hs
-- 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
-- 
-- What is the smallest number that is evenly divisible by all of the numbers from 1 to 20?

my_gcd a b 
  | a < b = my_gcd b a
  | b == 0 = a
  | b > 0 = my_gcd b (a `mod` b)

my_lcm a b = (a*b) `div` (my_gcd a b)

my_lcm_list (n:ns) = foldl my_lcm n ns
    
numbers_up_to n
  | n <= 0 = []
  | n >= 1 = (numbers_up_to (n-1)) ++ [n]
  
result = my_lcm_list(numbers_up_to 20)